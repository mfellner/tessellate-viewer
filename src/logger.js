// @flow

import debug from 'debug';

import type Debugger from 'debug';

const PREFIX = 'tessellate-viewer';

type Logger = {
  debug: Debugger,
  info: Debugger,
  warn: Debugger,
  error: Debugger
};

export default function logger(name: string): Logger {
  return {
    debug: debug(`${PREFIX}:debug:${name}`),
    info: debug(`${PREFIX}:info:${name}`),
    warn: debug(`${PREFIX}:warn:${name}`),
    error: debug(`${PREFIX}:error:${name}`)
  };
}
