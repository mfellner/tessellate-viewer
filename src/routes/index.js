// @flow

import Router from 'koa-router';
import * as health from './health';
import * as view from './view';

export type Route = (router: Router) => Router;
export type RouteSet = { [key: string]: Route };

function values<T>(object: { [key: string]: T }): Array<T> {
  return Object.keys(object).map(key => object[key]);
}

function initializeRoutes(router: Router, ...routes: Array<RouteSet>): Router {
  return routes
    .reduce((routes, set) => routes.concat(values(set)), [])
    .reduce((router, route) => route(router), router);
}

const router = initializeRoutes(new Router(), health, view);

export default router.routes();
