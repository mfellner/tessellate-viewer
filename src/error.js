// @flow

import logger from './logger';

const log = logger('error');

export class Problem extends Error {
  title: string;
  detail: ?string;
  status: ?number;

  constructor(args: { title: string, detail?: string, status?: number }) {
    super(args.title);
    this.title = args.title;
    this.detail = args.detail;
    this.status = args.status;
  }
}

export default (async function error(ctx: Object, next: () => Promise<any>): Promise<any> {
  try {
    return await next();
  } catch (err) {
    log.error(err);
    ctx.status = err.status || err.code || 500;
    ctx.body = {
      title: err.title || err.message,
      detail: err.detail,
      status: ctx.status
    };
  }
});
