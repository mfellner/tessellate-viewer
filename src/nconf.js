// @flow

import fs from 'fs';
import path from 'path';
import nconf from 'nconf';
import yaml from 'js-yaml';
import logger from './logger';

const log = logger('nconf');

function readYamlFile(file: string): Object {
  try {
    return yaml.safeLoad(fs.readFileSync(file, 'utf8'));
  } catch (e) {
    log.warn(`Unable to load ${file}`);
    return {};
  }
}

const configFile = path.resolve(process.cwd(), 'config.yaml');

export default nconf
  .use('memory')
  .argv()
  .env()
  .add('config', { type: 'literal', store: readYamlFile(configFile) })
  .defaults({
    APP_PORT: 3003,
    MORGAN_FORMAT: 'dev',
    MORGAN_THRESHOLD: 0,
    FRAGMENT_SOURCE: 'http://localhost:3002/fragment'
  });
