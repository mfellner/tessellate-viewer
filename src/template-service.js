// @flow

import path from 'path';
import ejs from 'ejs';
import thenify from 'thenify';
import logger from './logger';

type FragmentType = {|
  name: string,
  src: string,
  css?: string,
  html: string
|};

const renderFile = thenify(ejs.renderFile.bind(ejs));
const log = logger('template-service');

// async function loadTemplateFile() {
//   path.resolve(__dirname, '../static/template.ejs')
//   log.debug('Using template file %s', templateFile)
//   return require('../static/template.ejs')
// }

export async function renderView(fragment: FragmentType): Promise<string> {
  log.debug('Render view...');
  const meta = {};
  const templateFile = path.resolve(__dirname, '../static/template.ejs');
  // $ FlowIssue
  // const templateFile = require('../static/template.ejs')
  log.debug('Using __dirname %s', __dirname);
  log.debug('Using template file %s', templateFile);
  return await renderFile(templateFile, { meta, fragment }, {});
}
