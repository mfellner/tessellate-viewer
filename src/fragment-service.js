// @flow

import request from 'request-promise-native';
import nconf from './nconf';
import logger from './logger';
import { Problem } from './error';

type FragmentResult = {|
  html: string,
  link: {|
    js: string,
    css?: string
  |}
|};

const log = logger('fragment-service');

class FragmentProblem extends Problem {
  constructor(detail: string) {
    super({ title: 'Fragment error.', detail, status: 404 });
  }
}

export async function fetchFragment(domain: string, name: string): Promise<FragmentResult> {
  const url = nconf.get('FRAGMENT_SOURCE');
  log.debug('Fetch fragment %s %s/%s', url, domain, name);
  const { body, headers } = await request({
    url,
    method: 'GET',
    headers: {
      'x-zalando-request-uri': name,
      'x-zalando-request-host': domain
    },
    resolveWithFullResponse: true
  });

  return { html: body, link: parseLinkHeader(headers.link) };
}

function parseLinkHeader(header: string): {| js: string, css?: string |} {
  const jsMatch = header.trim().match(/<(\S+\.js)>;/);
  const cssMatch = header.trim().match(/<(\S+\.css)>;/);

  if (!jsMatch) throw new FragmentProblem(`No fragment script found in ${header}`);

  return { js: jsMatch[1], css: cssMatch ? cssMatch[1] : undefined };
}
