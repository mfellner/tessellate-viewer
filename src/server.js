// @flow

import path from 'path';
import Koa from 'koa';
import morgan from 'koa-morgan';
import koaStatic from 'koa-static';
import nconf from './nconf';
import logger from './logger';
import routes from './routes';
import error from './error';

const log = logger('server');

export function init(): Koa {
  const app = new Koa();
  const morganFormat = nconf.get('MORGAN_FORMAT');
  const morganSkip = (req, res) => res.statusCode < nconf.get('MORGAN_THRESHOLD');

  const staticDir = path.resolve(__dirname, '../static');
  log.info('Serving files from %s', staticDir);

  return app
    .use(morgan(morganFormat, { skip: morganSkip }))
    .use(error)
    .use(koaStatic(staticDir, { defer: false, gzip: true }))
    .use(routes);
}

async function start(port: number) {
  init().listen(port);
  log.info('listening on port %d', port);
}

export function main() {
  start(parseInt(nconf.get('APP_PORT'))).catch(log.error);
}

if (require.main === module) {
  main();
}
