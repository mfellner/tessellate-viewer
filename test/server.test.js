/**
 * Server API integration tests.
 */

jest.mock('../src/fragment-service', () => ({
  fetchFragment: jest.fn()
}));

import supertest from 'supertest';
import * as server from '../src/server';

describe('server', () => {
  let app, request;

  beforeEach(() => {
    app = server.init().listen();
    request = supertest.agent(app);
  });

  afterEach(() => app.close());

  describe('/health', () => {
    it('should return OK', async () => {
      await request
        .get('/health')
        .expect(200)
        .expect('OK');
    });
  });

  describe('/:domain/:name', () => {
    it('should return HTML', async () => {
      const fragmentService = require('../src/fragment-service');
      fragmentService.fetchFragment.mockImplementation(() =>
        Promise.resolve({
          html: '<h1>Hello, test!</h1>',
          link: {
            js: 'http://test.com/test/index.js',
            css: 'http://test.com/test/index.css'
          }
        })
      );

      const { text } = await request
        .get('/zalando.de/test')
        .expect(200)
        .expect('Content-Type', /html/);

      expect(text).toMatchSnapshot();
    });
  });
});
