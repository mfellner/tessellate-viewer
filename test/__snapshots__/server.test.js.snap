// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server /:domain/:name should return HTML 1`] = `
"<!doctype html>
<html lang=\\"\\">
    <head>
        <meta charset=\\"utf-8\\">
        <meta http-equiv=\\"x-ua-compatible\\" content=\\"ie=edge\\">
        <title></title>
        <meta name=\\"description\\" content=\\"\\">
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">

        
        <link rel=\\"stylesheet\\" href=\\"http://test.com/test/index.css\\">
        

        <script data-main=\\"/js/main\\" src=\\"https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.2/require.min.js\\" integrity=\\"sha256-Vjusm6Kh2U7/tb6jBh+MOfxnaf2TWsTph34bMKhC1Qc=\\" crossorigin=\\"anonymous\\" defer></script>
    </head>
    <body>
      <fragment id=\\"test\\" src=\\"http://test.com/test/index.js\\">
        <h1>Hello, test!</h1>
      </fragment>
    </body>
</html>
"
`;
